internal/jobobject/limits.go

```diff
@@ -0,0 +1,266 @@
+	"github.com/pkg/errors"
+	"golang.org/x/sys/windows"
+)
+
+const (
+	processorWeightMax float64 = 10000
+
+	// MemoryLimitMax is the max memory limit a job object can be assigned
+	MemoryLimitMax uint64 = 0xffffffffffffffff
+)
+
+func isFlagSet(flag uint32, controlFlags uint32) bool {
+	if (flag & controlFlags) == flag {
+		return true
+	}
+	return false
+}
+
+// SetMemoryLimit sets the memory limit the job object represented by `jobHandle`
+// based on the given `memoryLimitInBytes`
+func SetMemoryLimit(jobHandle windows.Handle, memoryLimitInBytes uint64) error {
+	info, err := getExtendedInformation(jobHandle)
+	if err != nil {
+		return err
+	}
+
+	info.JobMemoryLimit = uintptr(memoryLimitInBytes)
+	info.BasicLimitInformation.LimitFlags |= windows.JOB_OBJECT_LIMIT_JOB_MEMORY
+	return setExtendedInformation(jobHandle, info)
+}
+
+func getExtendedInformation(jobHandle windows.Handle) (*windows.JOBOBJECT_EXTENDED_LIMIT_INFORMATION, error) {
+	info := windows.JOBOBJECT_EXTENDED_LIMIT_INFORMATION{}
+	if err := winapi.QueryInformationJobObject(
+		jobHandle,
+		windows.JobObjectExtendedLimitInformation,
+		uintptr(unsafe.Pointer(&info)),
+		uint32(unsafe.Sizeof(info)),
+		nil,
+	); err != nil {
+		return nil, err
+	}
+	return &info, nil
+}
+
+// GetMemoryLimit gets the memory limit in bytes of the job object represented by `jobHandle`
+func GetMemoryLimit(jobHandle windows.Handle) (uint64, error) {
+	info, err := getExtendedInformation(jobHandle)
+	if err != nil {
+		return 0, err
+	}
+
+	return uint64(info.JobMemoryLimit), nil
+}
+
+// SetCPULimit sets the CPU limit depending on the specified `CPURateControlType` to `rateControlValue` for the
+// job object represented by `jobHandle`
+func SetCPULimit(jobHandle windows.Handle, rateControlType CPURateControlType, rateControlValue uint32) error {
+	cpuInfo, err := getCPURateControlInformation(jobHandle)
+	if err != nil {
+		return err
+	}
+	switch rateControlType {
+	case WeightBased:
+		if rateControlValue < CPUWeightMin || rateControlValue > CPUWeightMax {
+			return fmt.Errorf("processor weight value of `%d` is invalid", rateControlValue)
+		}
+		cpuInfo.ControlFlags |= winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_WEIGHT_BASED
+		cpuInfo.Value = rateControlValue
+	case RateBased:
+		if rateControlValue < CPULimitMin || rateControlValue > CPULimitMax {
+			return fmt.Errorf("processor rate of `%d` is invalid", rateControlValue)
+		}
+		cpuInfo.ControlFlags |= winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP
+		cpuInfo.Value = rateControlValue
+	default:
+		return errors.New("invalid job object cpu rate control type")
+	}
+
+	if _, err := windows.SetInformationJobObject(
+		jobHandle,
+		windows.JobObjectCpuRateControlInformation,
+		uintptr(unsafe.Pointer(cpuInfo)),
+		uint32(unsafe.Sizeof(cpuInfo)),
+	); err != nil {
+		return fmt.Errorf("failed to set cpu limit info on job object: %s", err)
```

Comment:
This is a copy of my code I believe but I'd been meaning to swap all the uses of `fmt.Errorf` with `errors.Wrap` if we're just printing the error as a string at the end.

Reply:
I can make that change